demo/site/src/middleware/redirectToMainHost.ts

```diff
@@ -1,14 +1,61 @@
+import { gql } from "@comet/site-nextjs";
+import { type ExternalLinkBlockData, type InternalLinkBlockData, type RedirectsLinkBlockData } from "@src/blocks.generated";
+import { type GQLPageTreeNodeScope } from "@src/graphql.generated";
 import type { PublicSiteConfig } from "@src/site-configs";
+import { createSitePath } from "@src/util/createSitePath";
+import { createGraphQLFetch } from "@src/util/graphQLClient";
 import { getHostByHeaders, getSiteConfigForHost, getSiteConfigs } from "@src/util/siteConfig";
 import { type NextRequest, NextResponse } from "next/server";
 
+import { memoryCache } from "./cache";
 import { type CustomMiddleware } from "./chain";
 
-const normalizeDomain = (host: string) => (host.startsWith("www.") ? host.substring(4) : host);
+const domainRedirectsQuery = gql`
+    query DomainRedirects($scope: RedirectScopeInput!) {
+        paginatedRedirects(scope: $scope) {
+            nodes {
+                id
+                source
+                target
+                sourceType
+            }
+        }
+    }
+`;
+
+async function fetchDomainRedirects(domain: string) {
+    const key = `domainRedirects-${domain}`;
+    return memoryCache.wrap(key, async () => {
+        const graphQLFetch = createGraphQLFetch();
+        const data = await graphQLFetch<
+            { paginatedRedirects: { nodes: { id: string; source: string; target: RedirectsLinkBlockData; sourceType: string }[] } },
+            { scope: { domain: string } }
+        >(domainRedirectsQuery, {
+            scope: { domain },
+        });
+
+        return data?.paginatedRedirects?.nodes || [];
+    });
+}
+
+async function getDomainRedirectTarget(domain: string, host: string): Promise<RedirectsLinkBlockData | undefined> {
+    const redirects = await fetchDomainRedirects(domain);
+    const redirectsArray = Array.isArray(redirects) ? redirects : [redirects];
+    const normalizeHost = (value: string) => {
+        return value.replace(/^https?:\/\//, "");
+    };
+    const matching = redirectsArray.find((redirect) => {
+        return redirect.sourceType === "domain" && normalizeHost(redirect.source) === normalizeHost(host);
+    });
+    if (matching) {
+        return matching.target;
+    }
+    return undefined;
+}
```

Comment:
Just randomly saw this: if we extract the matching into a sync/pure helper that takes redirects + host, it gets way more testable.

```suggestion
async function getDomainRedirectTarget(redirects: whatever-type-it-is, host: string): Promise<RedirectsLinkBlockData | undefined> {
    const redirectsArray = Array.isArray(redirects) ? redirects : [redirects];
    const normalizeHost = (value: string) => { // this could be also extracted in a dedicated function
        return value.replace(/^https?:\/\//, "");
    };
    const matching = redirectsArray.find((redirect) => {
        return redirect.sourceType === "domain" && normalizeHost(redirect.source) === normalizeHost(host);
    });
    if (matching) {
        return matching.target;
    }
    return undefined;
}
```
That way we would achieve:
- Single responsibility → after the changes `getDomainRedirectTarget` is only responsible of normalizing and match the host. If we go even further we could extract the host normalization in separate function so we really achieve single responsability here.   
- This is much easier to test → unit tests don’t need mocks/stubs; you just pass a small redirects array and assert the returned target.